### ISSUE TITLE
The union type cannot be determined using 'string []' to determine the type

### ISSUE THREAD

**Issue by crazyair** [microsoft/TypeScript#55246]
### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/zh/play?#code/C4TwDgpgBAKuEEYoF4oG8oGdgCcCWAdgOYD8AXFAOQCGlA3FNTjtSOVDhNQCYD2BAGxBQA2jUoBdBkQjA4kdgApu1YNQriAlCgB8jAsIC+dAFChIseACYU6LLkKkNAI3qNmrdpx78hoyq5SUDJy8Eoqai6U2sh61AZQxiYmEAAeYLw4wFAAxvzYUOaIFPKIUAA+lpA2qBjY+MQatAA0wbKlFBHUulBdiQwA9ANQAETUI8lpGVm5+dlFViXwSJWlNXZMLCAUYrQSjJizBNitIR29qt2xF2r9UEP6ICZAA)

### 💻 Code

```ts
type Type1 = { string?: 'a'; array?: readonly ['a']; getType?: (data: 'a') => any };
type Type2 = { string?: 'b'; array?: readonly ['b']; getType?: (data: 'b') => any };

export const type1: Type1 | Type2 = { string: 'a', getType: data => data }; // "a"

export const type2: Type1 | Type2 = { array: ['a'] as const, getType: data => data }; // any

```

### 🙁 Actual behavior

input `['a']`, `getType` `data` is`any`

### 🙂 Expected behavior

input `['a']`, `getType` `data` is`a`

**Comment by crazyair** [comment]
I used const readonly, but still couldn't get it

**Comment by jcalz** [comment]
Please properly fill out the Version & Regression Information section.

Tuples aren't treated as discriminants; this is essentially a duplicate of #48500.

**Comment by crazyair** [comment]
Can we release the first level of derivation?

**Comment by crazyair** [comment]
Add keywords

**Comment by jcalz** [comment]
If I understand you correctly (and that's a big *if*, there seems to be a language barrier here), what you're asking for would require #18758

**Comment by jcalz** [comment]
"This changed between versions 4.9 and 5.1" <-- this means that it was working as you desired in version 4.9 and is broken in version 5.1.  Is that really what you mean?  I'm going to disengage now because it seems like there's a communication problem and I don't know how to fix it.  Good luck.

**Comment by crazyair** [comment]
Can it only support `(string | number)[]`

**Comment by Khayoon** [comment]
TypeScript doesn't natively support `(string | number)[]` as a discriminant in union types.

Try to restructure the code to make use of discriminant properties that TS can recognize. Here's something that might do the trick:

`type TypeA = { kind: 'a', array: ['a'], getType: (data: 'a') => any };`
`type TypeB = { kind: 'b', array: ['b'], getType: (data: 'b') => any };`
`type Type = TypeA | TypeB;`
`function processType(type: Type) {`
`    if (type.kind === 'a') {`
`        // TypeScript now knows 'type' is of type TypeA and 'type.getType' expects 'a'`
`        console.log(type.getType('a'));`
`    } else if (type.kind === 'b') {`
`        // TypeScript now knows 'type' is of type TypeB and 'type.getType' expects 'b'`
`        console.log(type.getType('b'));`
`    }`
`}`
`let type1: Type = { kind: 'a', array: ['a'], getType: data => data }; `
`let type2: Type = { kind: 'b', array: ['b'], getType: data => data }; `
`processType(type1); // "a"`
`processType(type2); // "b"`

 We're gonna play a little trick on TypeScript. We're gonna add this new thing called 'kind' to our TypeA and TypeB.

Now, these 'kind' fellas, they ain't just hanging around for the good times, they're our inside men! They let TypeScript in on the secret, you see? Are we dealing with a TypeA? Or maybe a TypeB? TypeScript knows, 'cause our 'kind' variable tipped it off!

Now, once TypeScript's got the lay of the land, it can figure out what's what. It knows if 'getType' is expecting an 'a' or a 'b', cause it knows if it's dealing with a TypeA or a TypeB, thanks to kind.

If we can't change how our objects look, if we can't get 'kind' through, we're back to square one, and TypeScript isn't gonna be any wiser about those tuple and array discriminants. So remember, this is by no means a perfect workaround, but it's a pretty darn good one if you can swing it!

**Comment by Khayoon** [comment]
TypeScript doesn't natively support `(string | number)[]` as a discriminant in union types.

Try to restructure the code to make use of discriminant properties that TS can recognize. Here's something that might do the trick:

`type TypeA = { kind: 'a', array: ['a'], getType: (data: 'a') => any };`
`type TypeB = { kind: 'b', array: ['b'], getType: (data: 'b') => any };`
`type Type = TypeA | TypeB;`
`function processType(type: Type) {`
`    if (type.kind === 'a') {`
`        // TypeScript now knows 'type' is of type TypeA and 'type.getType' expects 'a'`
`        console.log(type.getType('a'));`
`    } else if (type.kind === 'b') {`
`        // TypeScript now knows 'type' is of type TypeB and 'type.getType' expects 'b'`
`        console.log(type.getType('b'));`
`    }`
`}`
`let type1: Type = { kind: 'a', array: ['a'], getType: data => data }; `
`let type2: Type = { kind: 'b', array: ['b'], getType: data => data }; `
`processType(type1); // "a"`
`processType(type2); // "b"`

 We're gonna play a little trick on TypeScript. We're gonna add this new thing called 'kind' to our TypeA and TypeB.

Now, these 'kind' fellas, they ain't just hanging around for the good times, they're our inside men! They let TypeScript in on the secret, you see? Are we dealing with a TypeA? Or maybe a TypeB? TypeScript knows, 'cause our 'kind' variable tipped it off!

Now, once TypeScript's got the lay of the land, it can figure out what's what. It knows if 'getType' is expecting an 'a' or a 'b', cause it knows if it's dealing with a TypeA or a TypeB, thanks to kind.

If we can't change how our objects look, if we can't get 'kind' through, we're back to square one, and TypeScript isn't gonna be any wiser about those tuple and array discriminants. This is by no means a perfect workaround

**Comment by Khayoon** [comment]
The fact of the matter is, ts doesn't have built-in support for using tuples as discriminants in union types.

Instead of trying to use the 'array' property as a discriminant, you can introduce a new property, say, 'kind', which can be used as a string literal type discriminant.

Consider the following:

```
typescript
type Type1 = { kind: 'Type1', string?: 'a'; array?: readonly ['a']; getType?: (data: 'a') => any };
type Type2 = { kind: 'Type2', string?: 'b'; array?: readonly ['b']; getType?: (data: 'b') => any };

export const type1: Type1 | Type2 = { kind: 'Type1', string: 'a', getType: data => data }; // "a"
export const type2: Type1 | Type2 = { kind: 'Type2', array: ['a'] as const, getType: data => data }; // any

function getType(data: Type1 | Type2) {
    if (data.kind === 'Type1') {
        console.log(data.getType('a'));
    } else if (data.kind === 'Type2') {
        console.log(data.getType('b'));
    }
}

getType(type1); // "a"
getType(type2); // "b"
```
In this setup, the 'kind' property serves as the discriminant that allows TypeScript to infer whether the 'getType' function expects 'a' or 'b'. This approach takes advantage of a powerful TypeScript feature known as "discriminated unions". gl

**Comment by RyanCavanaugh** [comment]
A reference-typed discriminant is difficult to reason about and would break a lot of assumptions that CFA uses. These just aren't supported and are unlikely to be supported in the future.

**Comment by crazyair** [comment]
The initial requirement was
```ts
interface TableProps<T = any> {
  // dataIndex support `a | b | [a] | [b]`
  columns: { dataIndex: keyof T | [keyof T]; render: (value: T[keyof T]) => any }[];
}

export const data: TableProps<{ a: string; b: number }> = {
  columns: [
    { dataIndex: ['a'], render: value => value }, // render value type is string
    { dataIndex: ['b'], render: value => value }, // render value type is number
  ],
};

```

**Comment by crazyair** [comment]
Can I convert `['a' , 'b']` to `a-b`
